### PR TITLE
[driver] Add MMC5603NJ Compass driver

### PR DESCRIPTION
**Pull request by salkinium**
1. Adds a driver for the cheapest and smallest compass on [Digikey: the MMC5603NJ](https://www.digikey.de/product-detail/de/memsic-inc/MMC5603NJ/1267-MMC5603NJCT-ND/10452803).
2. Makes `modm::ui::Animation` virtual, so that it can be used for motor position animation.
3. Fixes an issue with `Apa102::getColor()`

### CODE REVIEW

File: src/modm/driver/inertial/mmc5603.hpp [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+// forward declaration for friending with mmc5603::Data
+template < class I2cMaster >
+class Mmc5603;
+
+/// @ingroup modm_driver_mmc5603
+struct mmc5603
+{
+	static constexpr
+	uint8_t addr(uint8_t factory=0)
+	{ return 0b0110000 | (factory & 0b111); }
```
Comment (rleh) [inline src/modm/driver/inertial/mmc5603.hpp:32]: Would you like to add an explanation as a comment?

("The MEMSIC device 7-bit device address is `0110000` where the three LSB’s are pre-programmed into the MMC5603NJ by the factory.")

File: src/modm/driver/inertial/mmc5603.hpp [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+	};
+	MODM_FLAGS8(Control1);
+
+	enum class
+	Bandwidth : uint8_t
+	{
+		Ms6_6 = 0x00,
+		Ms3_5 = 0x01,
+		Ms2_0 = 0x02,
+		Ms1_2 = 0x03,
```
Comment (rleh) [inline src/modm/driver/inertial/mmc5603.hpp]:
```suggestion
		Time6600us = 0x00,
		Time3500us = 0x01,
		Time2000us = 0x02,
		Time1200us = 0x03,
```

... could be a little bit more self explaining.

File: src/modm/driver/inertial/mmc5603.hpp [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+		float inline
+		y() const { return swapData(1); }
+
+		float inline
+		z() const { return swapData(2); }
+		///@}
+
+		/// returns the temperature in Celcius
+		int8_t inline
+		t() const { return int16_t(data[9]) * 4/5 - 75; }
```
Comment (rleh) [inline src/modm/driver/inertial/mmc5603.hpp]:
```suggestion
		getTemperature() const { return int16_t(data[9]) * 4/5 - 75; }
```

or

```suggestion
		temperature() const { return int16_t(data[9]) * 4/5 - 75; }
```

File: src/modm/driver/inertial/mmc5603.hpp [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+	{ return update(Register::InternalControl0, Control0::Take_meas_T); }
+
+	modm::ResumableResult<bool>
+	configureContinuousMode(uint8_t frequency)
+	{
+		RF_BEGIN();
+		if (not RF_CALL(write(Register::ODR, frequency)))
+			RF_RETURN(false);
+		if (not RF_CALL(update(Register::InternalControl1, Register_t(i(Bandwidth::Ms1_2)))))
+		    RF_RETURN(false);
```
Comment (rleh) [inline src/modm/driver/inertial/mmc5603.hpp]: Wouldn't it make sense to be able to configure the bandwidth if the measurement frequency is configurable?

```suggestion
	modm::ResumableResult<bool>
	configureContinuousMode(uint8_t frequency, Bandwidth bandwidth = Bandwidth::Ms1_2)
	{
		RF_BEGIN();
		if (not RF_CALL(write(Register::ODR, frequency)))
			RF_RETURN(false);
		if (not RF_CALL(update(Register::InternalControl1, Register_t(i(bandwidth)))))
		    RF_RETURN(false);
```

File: src/modm/driver/inertial/mmc5603.lb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+def init(module):
+    module.name = ":driver:mmc5603"
+    module.description = """
+# MMC5603 3-Axis Digital Magnetometer
+
+The MMC5603NJ is a monolithic complete 3-axis AMR magnetic sensor with on-chip
+signal processing and integrated I2C bus.
+
+It can measure magnetic fields within the full scale range of ±30 Gauss (G),
+with up to 0.0625mG per LSB resolution at 20bits operation mode and 2mG total
```
Comment (rleh) [inline src/modm/driver/inertial/mmc5603.lb:23]: But this driver only used the 16bit mode, correct?

**Review by rleh — CHANGES_REQUESTED** [review]

File: src/modm/ui/animation/base.hpp [hunk 6]
```diff
@@ -23,6 +23,39 @@ namespace modm
...
+	///			`false` if otherwise
+	virtual bool
+	animateTo(T value, uint32_t time) = 0;
+
+	/// @return	`true` if the value has been changed,
+	///			`false` otherwise
+	virtual bool
+	update()
+	{ return false; }
+};
```
Comment (chris-durand) [inline src/modm/ui/animation/base.hpp:60]: This base class should either have a public virtual or protected non-virtual destructor, see http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-dtor-virtual.

**Review by chris-durand — CHANGES_REQUESTED** [review]

File: src/modm/driver/inertial/mmc5603.lb (hunk 5 above)
Comment (salkinium) [reply src/modm/driver/inertial/mmc5603.lb:23]: No the driver reads all 20-bits, but the Data object only used 16. I change this to use all 20 now.

File: src/modm/driver/inertial/mmc5603.hpp (hunk 2 above)
Comment (salkinium) [reply src/modm/driver/inertial/mmc5603.hpp]: Added some docs from the datasheet.

File: src/modm/driver/inertial/mmc5603.lb (hunk 5 above)
Comment (rleh) [reply src/modm/driver/inertial/mmc5603.lb:23]: I don't know if that would be useful, so do what works best for you.

**Review by rleh — APPROVED** [review]

File: src/modm/driver/inertial/mmc5603.lb (hunk 5 above)
Comment (salkinium) [reply src/modm/driver/inertial/mmc5603.lb:23]: I mean the data is transferred anyways , might as well access them for a few instructions more.
I made the expensive float optional, so the 20-bit int32_t output is going to be cheap…

File: src/modm/driver/inertial/mmc5603.hpp (hunk 4 above)
Comment (salkinium) [reply src/modm/driver/inertial/mmc5603.hpp]: Yes, also the 1.2ms measurement time was only necessary for 1000Hz operation, however, that does not work with Automatic Set/Reset as enabled in the function, hence the frequency only being 8-bit for a max of 255Hz. If you need 1kHz update rate, then you need to do it manually or update the driver later.

**Comment by salkinium** [comment]
Still have to do:
- [x] Test the changes to `mmc5603::Data` class in hardware.

**Pull request merged**